File: suspension_model/suspension_elements/tire.py
from suspension_model.assets.misc_linalg import rotation_matrix
from suspension_model.suspension_elements.kingpin import Kingpin
from suspension_model.suspension_elements.node import Node
import numpy as np


class Tire:

    def __init__(self, 
                 contact_patch: Node, 
                 kingpin: Kingpin,  
                 static_gamma: float,
                 static_toe: float,
                 radius: float, 
                 width: float) -> None:
        
        self.cp: Node = contact_patch
        self.kingpin = kingpin
        self.gamma: float = static_gamma
        self.radius = radius
        self.width = width
        self.static_toe = static_toe
        self._induced_steer = static_toe

        # Calculate initial center
        rotation = rotation_matrix([1, 0, 0], self.gamma)
        self.initial_center = np.matmul(rotation, [0, 0, 1]) * self.radius + self.cp.position

        # Calculate center relative to kingpin
        ang_x, ang_y = self.kingpin.normalized_transform()
        center_shifted = self.initial_center - self.kingpin.inboard_node.position
        x_rot = rotation_matrix(unit_vec=[1, 0, 0], theta=ang_x)
        y_rot = rotation_matrix(unit_vec=[0, 1, 0], theta=-1 * ang_y)
        self.center_to_kingpin = np.matmul(y_rot, np.matmul(x_rot, center_shifted))

        # Calculate contact patch relative to kingpin
        ang_x, ang_y = self.kingpin.normalized_transform()
        cp_shifted = self.cp.initial_position - self.kingpin.inboard_node.position
        x_rot = rotation_matrix(unit_vec=[1, 0, 0], theta=ang_x)
        y_rot = rotation_matrix(unit_vec=[0, 1, 0], theta=-1 * ang_y)
        self.cp_to_kingpin = np.matmul(y_rot, np.matmul(x_rot, cp_shifted))

        # Calculate static direction
        rotation = rotation_matrix([1, 0, 0], self.gamma)
        self.initial_kpi, self.initial_caster = self.kingpin.normalized_transform()
        self.initial_direction = np.matmul(rotation, [0, 1, 0])
        
        # Calculate direction relative to kingpin
        ang_x, ang_y = self.kingpin.normalized_transform()
        x_rot = rotation_matrix(unit_vec=[1, 0, 0], theta=ang_x)
        y_rot = rotation_matrix(unit_vec=[0, 1, 0], theta=-1 * ang_y)
        self.tire_direction = np.matmul(y_rot, np.matmul(x_rot, self.initial_direction))

        self.elements = [self.cp]
    
    def plot_elements(self, plotter):
        plotter.add_tire(center=self.center, direction=self.direction, radius=self.radius, height=self.height)
        for element in self.elements:
            element.plot_elements(plotter=plotter)
    
    @property
    def direction(self):
        ang_x, ang_y = self.kingpin.normalized_transform()
        x_rot = rotation_matrix(unit_vec=[1, 0, 0], theta=-1 * ang_x)
        y_rot = rotation_matrix(unit_vec=[0, 1, 0], theta=ang_y)
        z_rot = rotation_matrix(unit_vec=[0, 0, 1], theta=self.induced_steer)
        direction = np.matmul(y_rot, np.matmul(x_rot, np.matmul(z_rot, self.tire_direction)))

        return direction

    @property
    def center(self):
        ang_x, ang_y = self.kingpin.normalized_transform()
        x_rot = rotation_matrix(unit_vec=[1, 0, 0], theta=-1 * ang_x)
        y_rot = rotation_matrix(unit_vec=[0, 1, 0], theta=ang_y)
        z_rot = rotation_matrix(unit_vec=[0, 0, 1], theta=self.induced_steer)
        center_position = np.matmul(y_rot, np.matmul(x_rot, np.matmul(z_rot, self.center_to_kingpin))) + self.kingpin.inboard_node.position

        return center_position

    @property
    def height(self):
        return self.width
    
    @property
    def induced_steer(self):
        return self._induced_steer
    
    @induced_steer.setter
    def induced_steer(self, value: float):
        self._induced_steer = self.static_toe
        self._induced_steer += value

        # Update contact patch position
        ang_x, ang_y = self.kingpin.normalized_transform()
        x_rot = rotation_matrix(unit_vec=[1, 0, 0], theta=-1 * ang_x)
        y_rot = rotation_matrix(unit_vec=[0, 1, 0], theta=ang_y)
        z_rot = rotation_matrix(unit_vec=[0, 0, 1], theta=self._induced_steer)
        self.cp.position = np.matmul(y_rot, np.matmul(x_rot, np.matmul(z_rot, self.cp_to_kingpin))) + self.kingpin.inboard_node.position

File: src/applications/kin_cpp/tertiary_elements/tire.cpp
#include "tire.h"
#include "../assets/misc_linalg.h"

Tire::Tire(Node *contact_patch, Kingpin *kp, double _static_gamma, double _static_toe, double _radius, double _width) {
    cp = contact_patch;
    kingpin = kp;
    gamma = _static_gamma;
    static_toe = _static_toe;
    radius = _radius;
    width = _width;
    _induced_steer = _static_toe;

    /* Calculate initial center */
    StaticMatrix<double, 3UL, 3UL> rotation = rotation_matrix({1, 0, 0}, gamma);
    StaticVector<double, 3UL> z_axis({0, 0, 1});
    initial_center = (rotation * z_axis) * radius + cp->position;

    /* Calculate center relative to kingpin */
    StaticVector<double, 2UL> xy_angles = kingpin->getBeam()->normalized_transform();
    StaticVector<double, 3UL> center_shifted = initial_center - kingpin->getBeam()->getInboardNode()->position;
    StaticMatrix<double, 3UL, 3UL> x_rot = rotation_matrix({1, 0, 0}, xy_angles[0]);
    StaticMatrix<double, 3UL, 3UL> y_rot = rotation_matrix({0, 1, 0}, -1 * xy_angles[1]);
    center_to_kingpin = y_rot * (x_rot * center_shifted);
    
    /* Calculate contact patch relative to kingpin */
    StaticVector<double, 3UL> cp_shifted = cp->initial_position - kingpin->getBeam()->getInboardNode()->position;
    cp_to_kingpin = y_rot * (x_rot * cp_shifted);

    /* Calculate static direction */
    initial_kpi = xy_angles[0];
    initial_caster = xy_angles[1];
    StaticVector<double, 3UL> y_axis({0, 1, 0});
    initial_direction = rotation * y_axis;

    /* Calculate direction relative to kingpin */
    tire_direction = y_rot * (x_rot * initial_direction);

    elements[0] = all_elements[0] = cp;
}

void Tire::translate (StaticVector<double, 3UL> &translation) {
    for (Node *element : all_elements) {
        element->translate(translation);
    }
}

void Tire::flatten_rotate(StaticVector<double, 3UL> &angle) {
    for (Node *element : all_elements) {
        element->flatten_rotate (angle);
    }
}

StaticVector<double, 3UL> Tire::direction() const {
    StaticVector<double, 2UL> xy_angles = kingpin->getBeam()->normalized_transform();
    StaticMatrix<double, 3UL, 3UL> x_rot = rotation_matrix({1, 0, 0}, -1 * xy_angles[0]);
    StaticMatrix<double, 3UL, 3UL> y_rot = rotation_matrix({0, 1, 0}, xy_angles[1]);
    StaticMatrix<double, 3UL, 3UL> z_rot = rotation_matrix({0, 0, 1}, _induced_steer);

    StaticVector<double, 3UL> direction = x_rot * (y_rot * (z_rot * tire_direction));
    return direction;
}

StaticVector<double, 3UL> Tire::center() const {
    StaticVector<double, 2UL> xy_angles = kingpin->getBeam()->normalized_transform();
    StaticMatrix<double, 3UL, 3UL> x_rot = rotation_matrix({1, 0, 0}, -1 * xy_angles[0]);
    StaticMatrix<double, 3UL, 3UL> y_rot = rotation_matrix({0, 1, 0}, xy_angles[1]);
    StaticMatrix<double, 3UL, 3UL> z_rot = rotation_matrix({0, 0, 1}, _induced_steer);

    StaticVector<double, 3UL> center_position = y_rot * (x_rot * (z_rot * center_to_kingpin)) + kingpin->getBeam()->getInboardNode()->position;

    return center_position;
}

double Tire::induced_steer() const {
    return _induced_steer;
}

void Tire::set_induced_steer(double value) {
    _induced_steer = static_toe;
    _induced_steer += value;

    StaticVector<double, 2UL> xy_angles = kingpin->getBeam()->normalized_transform();
    StaticMatrix<double, 3UL, 3UL> x_rot = rotation_matrix({1, 0, 0}, -1 * xy_angles[0]);
    StaticMatrix<double, 3UL, 3UL> y_rot = rotation_matrix({0, 1, 0}, xy_angles[1]);
    StaticMatrix<double, 3UL, 3UL> z_rot = rotation_matrix({0, 0, 1}, _induced_steer);

    cp->position = x_rot * (y_rot * (z_rot * cp_to_kingpin)) + kingpin->getBeam()->getInboardNode()->position;
}